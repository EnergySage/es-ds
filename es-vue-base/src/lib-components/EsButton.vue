<template>
    <b-button
        :class="{ 'inline': inline }"
        :variant="computedVariant"
        :size="size"
        v-bind="$attrs"
        v-on="$listeners">
        <!--
            @slot Button Content
            @binding {string} text or html of the button content
        -->
        <slot />
    </b-button>
</template>

<script lang="js">
import { BButton } from 'bootstrap-vue';

export default {
    name: 'EsButton',
    components: {
        BButton,
    },
    props: {
        /**
         * Button Variant
         *
         */
        variant: {
            type: String,
            default: 'primary',
            required: false,
        },
        /**
         * Outline only button
         */
        outline: {
            type: Boolean,
            default: false,
        },
        /**
         * Size of button
         */
        size: {
            type: String,
            validator: (val) => ['lg', 'md', 'sm'].includes(val),
            default: 'md',
        },
        /**
         * Only works for 'link' variant buttons.
         * Removes the fixed height and padding so the button can fit nicely within a block of text.
         */
        inline: {
            type: Boolean,
            default: false,
        },
    },
    computed: {
        /**
         * Compute the Variant
         * used for when you want to `outline` something but don't know the
         * theme's default variant
         *
         * @returns { string } computed variant based on outline prop
         */
        computedVariant() {
            return `${this.outline && this.variant !== 'link' ? 'outline-' : ''}${this.variant}`;
        },
    },
};
</script>

<style lang="scss" scoped>
.btn-link.inline {
    /* use normal CSS here so users can override with utility classes as necessary */
    border: 0;
    height: auto;
    line-height: inherit;
    margin: 0;
    padding: 0;
    vertical-align: baseline;
}
</style>
