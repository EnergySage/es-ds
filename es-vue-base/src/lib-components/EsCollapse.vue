<template>
    <div>
        <EsButton
            block
            :aria-expanded="expanded.toString()"
            :aria-controls="id"
            class="rounded-0 collapse-holder pb-0 p-0 text-left font-weight-bold text-black d-flex align-items-center justify-content-between text-decoration-none text-body"
            :class="{'border-top': border}"
            inline
            variant="link"
            @click="userClick">
            <div class="pt-100 pb-100">
                <slot name="title" />
            </div>
            <div>
                <IconChevronDown
                    :class="{
                        svg: true,
                        chevronExpanded: expanded,
                    }"
                    width="30px"
                    height="30px" />
            </div>
        </EsButton>

        <b-collapse
            :id="id"
            v-model="expanded"
            data-testid="collapse"
            v-on="$listeners">
            <slot />
        </b-collapse>

        <div
            :class="{
                'border-bottom': border,
                'pb-100': (expanded && !border), // should match title-slot top-padding
                expanded: expanded,
            }" />
    </div>
</template>

<script lang="js">
import { BCollapse } from 'bootstrap-vue';
import IconChevronDown from '../lib-icons/icon-chevron-down.vue';
import EsButton from './EsButton.vue';

export default {
    name: 'EsCollapse',
    components: {
        EsButton, BCollapse, IconChevronDown,
    },
    model: {
        prop: 'visible',
        event: 'userClick',
    },
    props: {
        /**
         * ID
         */
        id: {
            type: String,
            required: true,
        },
        /**
         * Visible
         * Suggested open/closed state. Will be ignored if and when the user interacts with the collapse.
         */
        visible: {
            type: Boolean,
            required: false,
            default: false,
        },
        /**
         * isProgrammaticUntilUserInput
         * Prioritize the visible prop over the user's interaction with the collapse.
         */
        isProgrammaticUntilUserInput: {
            type: Boolean,
            required: false,
            default: true,
        },
        border: {
            type: Boolean,
            required: false,
            default: true,
        },
    },
    data() {
        return {
            userSpecifiedIsExpanded: null,
            isExpanded: null, // We use a second variable to track the expanded state so that we only emit the toggled
            // event when the user interacts with the collapse.
        };
    },
    computed: {
        expanded: {
            get() {
                if (this.isProgrammaticUntilUserInput && this.userSpecifiedIsExpanded !== null) {
                    return this.userSpecifiedIsExpanded;
                }
                return this.visible;
            },
            set() {
                // noop
            },
        },
    },
    watch: {
        userSpecifiedIsExpanded(newValue) {
            this.isExpanded = newValue;
            this.$emit('toggled', newValue);
        },
        visible(newValue) {
            if (!this.isProgrammaticUntilUserInput) {
                this.isExpanded = newValue;
            }
        },
    },
    methods: {
        userClick() {
            this.userSpecifiedIsExpanded = !this.expanded;
            this.$emit('userClick', !this.expanded);
        },
    },
};
</script>

<style lang="scss" scoped>
.svg {
    transition: transform .5s ease;
}

.chevronExpanded {
    transform: rotate(180deg);
}

// .bottom-spacer {
//     padding-top: 2rem;

//     &.expanded {
//         padding-top: 3rem;
//     }
// }
</style>
