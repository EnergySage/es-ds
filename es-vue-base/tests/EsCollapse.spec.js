import { mount } from '@vue/test-utils';
import EsCollapse from '@/src/lib-components/EsCollapse.vue';
import EsButton from '@/src/lib-components/EsButton.vue';
import jestVue from '@/tests/jest.vue.config';

describe('EsCollapse', () => {
    // Basic test; should exist for most components
    test('<EsCollapse />', async () => {
        const wrapper = mount(EsCollapse, {
            ...jestVue,
            slots: {
                title: 'My Title',
                default: 'Lorem ipsum dolor sit amet, consectetur adipiscing.',
            },
            propsData: {
                id: 'myId',
            },
        });
        const a11y = await axe(wrapper.element);

        expect(wrapper.vm).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot();
        expect(a11y).toHaveNoViolations();
    });

    test('EsButton exists', () => {
        const wrapper = mount(EsCollapse, {
            ...jestVue,
            slots: {
                title: 'My Title',
                default: 'Lorem ipsum dolor sit amet, consectetur adipiscing.',
            },
            propsData: {
                id: 'myId',
            },
        });
        expect(wrapper.findComponent(EsButton).exists()).toBe(true);
        expect(wrapper.html()).toMatchSnapshot();
    });

    test('Title and default slot exist inside EsButton', async () => {
        const wrapper = mount(EsCollapse, {
            ...jestVue,
            slots: {
                title: 'My Title',
                default: 'Lorem ipsum dolor sit amet, consectetur adipiscing.',
            },
            propsData: {
                id: 'myId',
            },
        });
        const box = wrapper.findComponent(EsButton);
        await box.trigger('click');
        expect(box.text()).toContain('My Title');
        expect(wrapper.text()).toBe('My Title  Lorem ipsum dolor sit amet, consectetur adipiscing.');
        expect(wrapper.html()).toMatchSnapshot();
    });

    test('Icons switch upon button click', async () => {
        const wrapper = mount(EsCollapse, {
            ...jestVue,
            slots: {
                title: 'My Title',
                default: 'Lorem ipsum dolor sit amet, consectetur adipiscing.',
            },
            propsData: {
                id: 'myId',
            },
        });
        const box = wrapper.findComponent(EsButton);
        const content = wrapper.find('.collapse');

        expect(content.isVisible()).toBe(false);

        await box.trigger('click');
        // TODO: Check SVG changed class
        expect(content.isVisible()).toBe(true);
        expect(wrapper.html()).toMatchSnapshot();
    });
});
