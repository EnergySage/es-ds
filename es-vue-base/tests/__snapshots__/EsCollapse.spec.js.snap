// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EsCollapse <EsCollapse /> 1`] = `
"<div><button aria-label=\\"myId\\" type=\\"button\\" class=\\"btn collapse-holder text-left p-0 mb-3 font-weight-bold text-black d-flex align-items-center justify-content-between text-decoration-none text-body btn-link btn-md btn-block\\" block=\\"\\">
    <div>My Title</div>
    <div><svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 24 24\\" fill=\\"currentColor\\" style=\\"height: 30px; width: 30px;\\">
        <path fill-rule=\\"evenodd\\" clip-rule=\\"evenodd\\" d=\\"M3.293 8.293a1 1 0 0 1 1.414 0L12 15.586l7.293-7.293a1 1 0 1 1 1.414 1.414l-8 8a1 1 0 0 1-1.414 0l-8-8a1 1 0 0 1 0-1.414Z\\"></path>
      </svg></div>
  </button>
  <transition-stub css=\\"true\\" enterclass=\\"\\" leaveclass=\\"collapse show\\" entertoclass=\\"collapse show\\" leavetoclass=\\"collapse\\" enteractiveclass=\\"collapsing\\" leaveactiveclass=\\"collapsing\\" aria-labelledby=\\"myId\\" role=\\"tabpanel\\" data-testid=\\"collapse\\">
    <div id=\\"myId\\" class=\\"collapse\\" style=\\"display: none;\\">
      <div class=\\"pb-5\\">Lorem ipsum dolor sit amet, consectetur adipiscing.</div>
    </div>
  </transition-stub>
  <div class=\\"border-bottom\\"></div>
</div>"
`;

exports[`EsCollapse EsButton exists 1`] = `
"<div><button aria-label=\\"myId\\" type=\\"button\\" class=\\"btn collapse-holder text-left p-0 mb-3 font-weight-bold text-black d-flex align-items-center justify-content-between text-decoration-none text-body btn-link btn-md btn-block\\" block=\\"\\">
    <div>My Title</div>
    <div><svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 24 24\\" fill=\\"currentColor\\" style=\\"height: 30px; width: 30px;\\">
        <path fill-rule=\\"evenodd\\" clip-rule=\\"evenodd\\" d=\\"M3.293 8.293a1 1 0 0 1 1.414 0L12 15.586l7.293-7.293a1 1 0 1 1 1.414 1.414l-8 8a1 1 0 0 1-1.414 0l-8-8a1 1 0 0 1 0-1.414Z\\"></path>
      </svg></div>
  </button>
  <transition-stub css=\\"true\\" enterclass=\\"\\" leaveclass=\\"collapse show\\" entertoclass=\\"collapse show\\" leavetoclass=\\"collapse\\" enteractiveclass=\\"collapsing\\" leaveactiveclass=\\"collapsing\\" aria-labelledby=\\"myId\\" role=\\"tabpanel\\" data-testid=\\"collapse\\">
    <div id=\\"myId\\" class=\\"collapse\\" style=\\"display: none;\\">
      <div class=\\"pb-5\\">Lorem ipsum dolor sit amet, consectetur adipiscing.</div>
    </div>
  </transition-stub>
  <div class=\\"border-bottom\\"></div>
</div>"
`;

exports[`EsCollapse Icons switch upon button click 1`] = `
"<div><button aria-label=\\"myId\\" type=\\"button\\" class=\\"btn collapse-holder text-left p-0 mb-3 font-weight-bold text-black d-flex align-items-center justify-content-between text-decoration-none text-body btn-link btn-md btn-block\\" block=\\"\\">
    <div>My Title</div>
    <div><svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 24 24\\" fill=\\"currentColor\\" style=\\"height: 30px; width: 30px;\\">
        <path fill-rule=\\"evenodd\\" clip-rule=\\"evenodd\\" d=\\"M11.293 6.293a1 1 0 0 1 1.414 0l8 8a1 1 0 0 1-1.414 1.414L12 8.414l-7.293 7.293a1 1 0 0 1-1.414-1.414l8-8Z\\"></path>
      </svg></div>
  </button>
  <transition-stub css=\\"true\\" enterclass=\\"\\" leaveclass=\\"collapse show\\" entertoclass=\\"collapse show\\" leavetoclass=\\"collapse\\" enteractiveclass=\\"collapsing\\" leaveactiveclass=\\"collapsing\\" aria-labelledby=\\"myId\\" role=\\"tabpanel\\" data-testid=\\"collapse\\">
    <div id=\\"myId\\" class=\\"collapse show\\" style=\\"\\">
      <div class=\\"pb-5\\">Lorem ipsum dolor sit amet, consectetur adipiscing.</div>
    </div>
  </transition-stub>
  <div class=\\"border-bottom\\"></div>
</div>"
`;

exports[`EsCollapse Title and default slot exist inside EsButton 1`] = `
"<div><button aria-label=\\"myId\\" type=\\"button\\" class=\\"btn collapse-holder text-left p-0 mb-3 font-weight-bold text-black d-flex align-items-center justify-content-between text-decoration-none text-body btn-link btn-md btn-block\\" block=\\"\\">
    <div>My Title</div>
    <div><svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 24 24\\" fill=\\"currentColor\\" style=\\"height: 30px; width: 30px;\\">
        <path fill-rule=\\"evenodd\\" clip-rule=\\"evenodd\\" d=\\"M11.293 6.293a1 1 0 0 1 1.414 0l8 8a1 1 0 0 1-1.414 1.414L12 8.414l-7.293 7.293a1 1 0 0 1-1.414-1.414l8-8Z\\"></path>
      </svg></div>
  </button>
  <transition-stub css=\\"true\\" enterclass=\\"\\" leaveclass=\\"collapse show\\" entertoclass=\\"collapse show\\" leavetoclass=\\"collapse\\" enteractiveclass=\\"collapsing\\" leaveactiveclass=\\"collapsing\\" aria-labelledby=\\"myId\\" role=\\"tabpanel\\" data-testid=\\"collapse\\">
    <div id=\\"myId\\" class=\\"collapse show\\" style=\\"\\">
      <div class=\\"pb-5\\">Lorem ipsum dolor sit amet, consectetur adipiscing.</div>
    </div>
  </transition-stub>
  <div class=\\"border-bottom\\"></div>
</div>"
`;
